static-instrumenter/agent-instrumenter/build.gradle.kts

```diff
@@ -1,12 +1,120 @@
   runtimeOnly("org.slf4j:slf4j-simple")
+
+  // TODO: remove local jar once agent with hook is released
+  javaagent(files("libs/opentelemetry-javaagent.jar"))
+  // javaagent("io.opentelemetry.javaagent:opentelemetry-javaagent")
+
+  bootstrapLibs(project(":static-instrumenter:bootstrap"))
+  javaagentLibs(project(":static-instrumenter:agent-extension"))
+}
+
+// TODO: migrate to JVM test suite plugin
+testSets {
+  create("integrationTest")
+}
+
+tasks {
+
+  val relocateJavaagentLibs by registering(ShadowJar::class) {
+    configurations = listOf(javaagentLibs)
+    duplicatesStrategy = DuplicatesStrategy.FAIL
+    archiveFileName.set("javaagentLibs-relocated.jar")
+  }
+
+  shadowJar {
+    configurations = listOf(javaagent, bootstrapLibs)
+
+    dependsOn(relocateJavaagentLibs)
+    isolateClasses(relocateJavaagentLibs.get().outputs.files)
+    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+
+    manifest {
+      attributes.put("Main-Class", "io.opentelemetry.contrib.staticinstrumenter.agent.main.Main")
+      attributes.put("Agent-Class", "io.opentelemetry.contrib.staticinstrumenter.agent.OpenTelemetryStaticAgent")
+      attributes.put("Premain-Class", "io.opentelemetry.contrib.staticinstrumenter.agent.OpenTelemetryStaticAgent")
+      attributes.put("Can-Redefine-Classes", "true")
+      attributes.put("Can-Retransform-Classes", "true")
+      attributes.put("Implementation-Vendor", "OpenTelemetry")
+      attributes.put("Implementation-Version", "demo-${project.version}")
+    }
+  }
+
+  // TODO: the resulting jar contains both inst and unpacked inst,
+  //  but we need only the unpacked inst here
+  val createNoInstAgent by registering(ShadowJar::class) {
+
+    configurations = listOf(javaagent)
+    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+
+    archiveClassifier.set("no-inst")
+
+    inputs.files.forEach {
+      if (it.name.endsWith(".jar")) {
+        from(zipTree(it)) {
+          include("inst/")
+          eachFile {
+            // drop inst/
+            relativePath = RelativePath(true, *relativePath.segments.drop(1).toTypedArray())
+            rename("(.*)\\.classdata\$", "\$1.class")
+          }
+        }
+      }
+    }
+  }
+
+  assemble {
+    dependsOn(shadowJar, createNoInstAgent)
+  }
+
+  val integrationTest by existing(Test::class) {
+    dependsOn(assemble)
+    jvmArgumentProviders.add(AgentJarsProvider(shadowJar.flatMap { it.archiveFile }, createNoInstAgent.flatMap { it.archiveFile }))
+  }
+}
+
+fun CopySpec.isolateClasses(jars: Iterable<File>) {
+  jars.forEach {
+    from(zipTree(it)) {
+      into("inst")
+      rename("^(.*)\\.class\$", "\$1.classdata")
+      // Rename LICENSE file since it clashes with license dir on non-case sensitive FSs (i.e. Mac)
+      rename("""^LICENSE$""", "LICENSE.renamed")
+      exclude("META-INF/INDEX.LIST")
+      exclude("META-INF/*.DSA")
+      exclude("META-INF/*.SF")
+    }
+  }
+}
+
+tasks.withType<ShadowJar>().configureEach {
```

Comment:
Nit: this and the block from line 120 can be put inside the huge `tasks {}` block above